Added Country Map : New Visualization tools

ymatagne opened the pull request:
Hello,

For a UC, I need to put metrics on regions of a specific country.

I created this feature in a generic way to add the different countries we want to implement.

For now, I have booted the feature with France and Spain. The addition of a new country is simple.

This is a component with D3.js

Good reading :)

At your disposal for more information.

Commit:
feat(visualization): Create new visualization's tools that display country with geojson file. Currently only  france and Spain are loaded

Commit:
Merge remote-tracking branch 'origin/master' into feat_viz_country_map

Commit:
refacto(visualization): fix issues created by codeclimate

Commit:
refacto(visualization): fix issues created by travis for mysql databases

Commit:
refacto(visualization): fix issues created by codeclimate

Commit:
refacto(visualization): fix issues created by codeclimate

ymatagne committed:
Merge branch 'master' into feat_viz_country_map

xrmx commented:
Wondering if there's a package (or an api) who has the geojsons so we don't have to add them one by one.

ymatagne commented:
I don't know if package or api exist. I can create this but i don't know if it's good idea because, if user don't have network, user will don't have tools.

Commit:
refacto(visualization): fix issues created by codeclimate

ymatagne committed:
Merge branch 'master' into feat_viz_country_map

Commit:
refacto(visualization): fix issues created by codeclimate

sanshanxiashi commented:
Do I need to reload the database?

mistercrunch commented on superset/assets/javascripts/explorev2/stores/controls.jsx:
```diff
@@ -265,7 +265,17 @@ export const controls = {
+    label: 'Country Name Type',
+    default: 'France',
+    choices: [
+        ['name', 'Full name'],
```
This first option should not be a choice.

mistercrunch reviewed (changes requested):
FYI the longer term approach for Geo data will be in the direction of using mapbox and probably just mapping deck.gl

I feel like 2 countries isn't reaching critical mass to get this feature more useful than confusing to most. We should aim at top 10 developed countries perhaps.

Another thing that is unclear is how the data provided should match the geo json. For instance for France you have to know that your data should provide department numbers, maybe as a string (I'd have to read the code to know). 

So either in the docs or ideally in the tooltip for the control you'd specify how to provide data.

If not contributing "critical mass" of number of countries, it would be good to specify somewhere how to add a new country to the set.

Knowing that this could be used for different grains, (as in "USA States" and "USA Counties"), you may want to be clear for "France Departments" and whatever they call those for spain.

ymatagne committed:
Merge branch 'master' into feat_viz_country_map

Commit:
fix(visualization): fix issues

Commit:
doc(visualization): Init visualization documentation

Commit:
doc(visualization): init visualization documentation for country

ymatagne committed:
Merge branch 'master' into feat_viz_country_map

ymatagne commented:
Hi,

I have just corrected all the remarks.

I think the creation of the component can be a complement to deck.gl. It brings a static side to visualization.

I added a lot of new country maps.

I also add a whole new documentation page.

I changed the system. We go through IDs rather than department, region, or province codes.

I also fixed some bugs.

What do you think ?

ymatagne committed:
Merge branch 'master' into feat_viz_country_map

mistercrunch commented on superset/assets/backendSync.json:
```diff
@@ -2284,6 +2357,12 @@
       "hidden": true,
       "description": "The id of the active slice"
+    },
+    "cache_timeout": {
```
Wait that does not belong here. I think it's in `master` already. Bad merge?

ymatagne replied:
Yes I think.

mistercrunch commented on superset/assets/javascripts/explorev2/stores/controls.jsx:
```diff
@@ -265,7 +265,28 @@ export const controls = {
+    type: 'SelectControl',
+    label: 'Country Name',
+    default: 'France',
+    choices: [
```
```
[
 'Algeria',
 'Belgium',
  [...]
].map(s => [s, s])
```

ymatagne replied:
Indeed, it's prettier.

mistercrunch commented on superset/assets/javascripts/explorev2/stores/controls.jsx:
```diff
@@ -265,7 +265,28 @@ export const controls = {
+  select_country: {
+    type: 'SelectControl',
+    label: 'Country Name',
+    default: 'France',
```
Yeah why not! Vive la France!

ymatagne replied:
;)

mistercrunch commented on superset/assets/javascripts/explorev2/stores/visTypes.js:
```diff
@@ -571,7 +571,29 @@ const visTypes = {
+    controlOverrides: {
+      entity: {
+        label: 'ID of region/province/department',
+        description: 'It\'s id of your region/province/department in your table who should match with id of region/province/department in geojson file. (see documentation for list of id)',
```
If you use double quotes you don't have to escape the single quote

ymatagne replied:
no problem.

mistercrunch reviewed (changes requested):
Alright I think we're getting close. One concern is around the IDs, are those some sort of standard iso code or flat out arbitrary? Seems like for the US I'm most likely to have a 2 letter code for state in my dataset, or perhaps the full state name, and that I'd like to use either without bringing in the mapping into my database or some crazy `CASE WHEN` expression.

Also it seems like the `geojson` files might have too much resolution, for example the US file is 90+mb which is pretty prohibitive. Perhaps we need to use lower resolution in some cases. I haven't looked into the size of the other files.

Apart from that things are looking pretty good.

xrmx commented on superset/assets/javascripts/explorev2/stores/controls.jsx:
```diff
@@ -265,7 +265,28 @@ export const controls = {
+        ['Italy', 'Italy'],
+        ['Morocco', 'Morocco'],
+        ['Netherlands', 'Netherlands'],
+        ['Russian', 'Russian'],
```
russia

ymatagne replied:
I corrected the different file names

xrmx reviewed (commented):
Looks like you have dup geojson: russia vs russian and egypt vs egypte

xrmx commented:
The netherlands have two geojson too.

ymatagne commented:
Hi @xrmx @mistercrunch !

Oh my God ! Files are very big! I m not saw before push! 
I can reduce all files ! It's not a problem. 
But before reduce files, i have question. 

I can put the department number or the code of a region for each country. However, this is a manual action to be performed before the files are integrated into the SUPERSET application.

Current IDs are more or less arbitrary numbers. I think this is an identifier generated by og2r according to the order of display of the regions on the map.

I would also prefer a department code or a region code for these IDs. On the other hand, a small manual action is to be carried out. I can write it in the documentation.

What do you think?

mistercrunch commented:
For example, the world map supports many different ISO codes for country mapping, trying to cater to whatever standard data you might have.

Now for things like regions there should be an obvious standard code, like the number of the french departments, the USA state 2 letter code, USA zip codes, ... In some cases it can be nice to also allow to point to the label but as in `California` instead of `ca` or `Alsace` instead of the department number, but that's not a requirement here and now, and that stuff gets tricky/imperfect with case, unicode and accentuation...

Commit:
fix(visualization): implement iso 3166-2 for id of country

Commit:
fix(visualization): implement iso 3166-2 for id of country

Commit:
doc(visualization): init visualization documentation for country

Commit:
fix(visualization): implement iso 3166-2 for id of country

ymatagne committed:
Merge branch 'master' into feat_viz_country_map

ymatagne commented:
Hi !

Sorry for late answer ! So, i implemented a new mechanical based of ISO 3166-2.

ISO 3166-2 is part of the ISO 3166 standard published by the International Organization for Standardization (ISO), and defines codes for identifying the principal subdivisions (e.g., provinces or states) of all countries coded in ISO 3166-1

I think it's great solution because that work with all countries

Size of geojson are bellow 1 Mo

Documentation is up to date.

What do you think?

bbs2009 commented:
@ymatagne Greetings. It is possible to add Ukraine to the list of countries. Thank you.

Commit:
feat(country_map): add ukraine map in component

ymatagne commented:
@bbs2009  I added Ukraine's map.

mistercrunch reviewed (changes requested):
This is getting pretty neat.
I pulled your branch, loaded the examples, and the first example `Country Map` doesn't seem to work out the the box.
* let's make the title descriptive "France population by department in 2006" or something like that
* point to the right column and metric
* I'm expecting the colors to be representative of the population (?)

I think that the second example `% Map` points to the wrong datasource...
![img](http://g.recordit.co/G1zQOa7SRm.gif)

Commit:
fix(visualization): Test dont working and add color based on metrics

ymatagne committed:
Merge branch 'master' into feat_viz_country_map

ymatagne commented:
Hi @mistercrunch 

It's ok for example... sorry.

I added new feature that change region's color based on metrics too.

mistercrunch commented on superset/assets/visualizations/country_map.js:
```diff
@@ -0,0 +1,190 @@
+      (maxMetric - minMetric) / 2,
+      (maxMetric - minMetric) / 3,
+      minMetric])
+    .range(['#d7191c', '#e76818', '#f29e2e', '#f9d057', '#ffff8c', '#90eb9d']);
```
What about leveraging this?:
https://github.com/airbnb/superset/blob/master/superset/assets/javascripts/explorev2/stores/controls.jsx#L122
along with
https://github.com/airbnb/superset/blob/366ecefbaaf4cc31234cc981ebab7eb420efe019/superset/assets/javascripts/modules/colors.js#L69

ymatagne replied:
It's same !

mistercrunch commented on superset/assets/visualizations/country_map.js:
```diff
@@ -0,0 +1,190 @@
+require('./country_map.css');
```
We're moving towards ES6 and away from using `require`
`import './country_map.css'`

mistercrunch reviewed (changes requested):
Ok I'm thinking these are my last comments, last stretch!

Commit:
refacto(visualization): fix issues for es6 and color component

ymatagne commented:
Hi @mistercrunch 

I think it's finally good!

I had not seen the existence of the color component. We agree, it was the same thing!

It's ok for you ?

mistercrunch commented:
Awesome PR, solid work @ymatagne !

ymatagne commented:
@mistercrunch  Thanks !
